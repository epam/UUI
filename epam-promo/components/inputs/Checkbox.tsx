import css from './Checkbox.scss';
import styles from '../../assets/styles/colorvars/inputs/checkbox-colorvars.scss';
import { Checkbox as uuiCheckbox, CheckboxProps } from '@epam/uui-components';
import { withMods } from '@epam/uui-core';
import { ReactComponent as Check_12 } from '../../icons/check-12.svg';
import { ReactComponent as Check_18 } from '../../icons/check-18.svg';
import { ReactComponent as PartlySelect_12 } from '../../icons/partly-select-12.svg';
import { ReactComponent as PartlySelect_18 } from '../../icons/partly-select-18.svg';

export interface CheckboxMods {
    /** Checkbox size */
    size?: '12' | '18';
    /** Set to 'dark' to make component work on a dark background */
    theme?: 'light' | 'dark';
    mode?: 'form' | 'cell';
}

export function applyCheckboxMods(mods: CheckboxMods & CheckboxProps) {
    return [css.root, css['size-' + (mods.size || '18')], css['mode-' + (mods.mode || 'form')], styles['checkbox-color-blue']];
}

export const Checkbox = withMods<CheckboxProps, CheckboxMods>(uuiCheckbox, applyCheckboxMods, props => ({
    icon: props.size === '12' ? Check_12 : Check_18,
    indeterminateIcon: props.size === '12' ? PartlySelect_12 : PartlySelect_18,
}));
