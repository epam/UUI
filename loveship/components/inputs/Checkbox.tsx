import * as types from '../types';
import css from './Checkbox.scss';
import { Checkbox as uuiCheckbox, CheckboxProps } from '@epam/uui-components';
import { withMods } from '@epam/uui-core';
import { ReactComponent as TickIcon } from '../icons/checkbox-checked.svg';
import { ReactComponent as IndeterminateIcon } from '../icons/checkbox-partial.svg';
import styles from '../../assets/styles/scss/loveship-color-vars.scss';

export interface CheckboxMods extends types.ColorMod {
    size?: '12' | '18';
    theme?: 'light' | 'dark';
    mode?: 'form' | 'cell';
}

export function applyCheckboxMods(mods: CheckboxMods & CheckboxProps) {
    return [
        css.root,
        css['size-' + (mods.size || '18')],
        css['theme-' + (mods.theme || 'light')],
        css['mode-' + (mods.mode || 'form')],
        styles['color-' + (mods.color || 'sky')],
    ];
}

export const Checkbox = withMods<CheckboxProps, CheckboxMods>(uuiCheckbox, applyCheckboxMods, () => ({
    icon: TickIcon,
    indeterminateIcon: IndeterminateIcon,
}));
