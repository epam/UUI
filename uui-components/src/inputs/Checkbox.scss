.container {
    display: flex;
    cursor: pointer;

    :global(.uui-checkbox) {
        display: flex;
        align-items: center;
        justify-content: center;
        cursor: pointer;
        flex-shrink: 0;

        input[type="checkbox"] {
            margin: 0;
            border: none;
            width: 0;
            height: 0;
            appearance: none;
            -webkit-appearance: none;
            -moz-appearance: none;


            &:focus, &:focus-visible {
                outline: none;
            }
        }
    }
}