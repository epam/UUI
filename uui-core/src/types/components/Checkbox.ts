import * as React from 'react';
import {
    ICanBeInvalid,
    ICanBeReadonly,
    ICheckable,
    IHasCX,
    IHasLabel,
    IAnalyticableOnChange,
    IHasRawProps,
    IHasForwardedRef,
    ICanFocus,
} from '../props';

export interface CheckboxCoreProps
    extends ICheckable,
        ICanFocus<HTMLInputElement>,
        IHasCX,
        ICanBeInvalid,
        IHasLabel,
        ICanBeReadonly,
        IAnalyticableOnChange<boolean>,
        IHasRawProps<React.LabelHTMLAttributes<HTMLLabelElement>>,
        IHasForwardedRef<HTMLLabelElement> {}
