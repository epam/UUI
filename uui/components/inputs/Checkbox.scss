@use  '../../assets/styles/index.scss' as *;

@mixin icon-sizes($size, $font-size, $line-height) {
    :global(.uui-checkbox) {
        width: $size;
        height: $size;

        :global(.uui-icon) {
            svg {
                width: $size;
            }
        }
    }

    :global(.uui-input-label) {
        font-size: $font-size;
        line-height: $line-height;
    }
}

.root {
    //Colors
    &:hover, &:active, &:focus-within {
        :global(.uui-checkbox) {
            border-color: var(--uui-checkbox-border-hover);
        }

        :global(.uui-checked) {
            background-color: var(--uui-checkbox-bg-checked-hover);
        }
    }

    :global(.uui-checkbox) {
        background-color: var(--uui-checkbox-bg);
        border: 1px solid var(--uui-checkbox-border);
        border-radius: var(--uui-checkbox-border-radius);
    }

    :global(.uui-input-label) {
        // @include typography-inline(); TODO: ???
        font-family: var(--font-regular);
        color: var(--uui-checkbox-label);
        margin-left: 12px;
        flex-grow: 1;
    }

    :global(.uui-checked) {
        background-color: var(--uui-checkbox-bg-checked);
        border-color: var(--uui-checkbox-border-checked);
        fill: var(--uui-checkbox-fill);
    }

    &:global(.uui-disabled),
    &:global(.uui-readonly) {
        cursor: default;

        :global(.uui-checkbox) {
            cursor: default;
            background-color: var(--uui-checkbox-bg-disabled);
            border-color: var(--uui-checkbox-border-disabled);
        }

        &:hover, &:global(.uui-checked), &:focus-within {
            &:global(.uui-checkbox) {
                border-color: var(--uui-checkbox-border-checked-disabled);
            }

            &:global(.uui-icon) {
                background-color: var(--uui-checkbox-bg-checked-disabled);
                border-color: var(--uui-checkbox-border-checked-disabled);
            }
        }

        :global(.uui-input-label) {
            color: var(--uui-checkbox-label-disabled);
        }

        :global(.uui-checked) {
            background-color: var(--uui-checkbox-bg-checked-disabled);
            border-color: var(--uui-checkbox-border-checked-disabled);
        }
    }

    &:global(.uui-invalid) {
        :global(.uui-checkbox) {
            border-color: var(--uui-checkbox-error);
            fill: var(--uui-checkbox-error);

            &:hover, &:active, &:focus-within {
                border-color: var(--uui-checkbox-error);
                fill: var(--uui-checkbox-error);
            }

            &:global(.uui-checked) {
                fill: var(--uui-checkbox-fill);
                background-color: var(--uui-checkbox-error);
                border-color: var(--uui-checkbox-error);

                &:hover, &:active, &:focus-within {
                    background-color: var(--uui-checkbox-error);
                    border-color: var(--uui-checkbox-error);
                }
            }
        }

        :global(.uui-input-label) {
            color: var(--uui-checkbox-error);
        }
    }

    //Sizes
    &.size-18 {
        @include icon-sizes(18px, 14px, 18px);
    }

    &.size-12 {
        @include icon-sizes(12px, 12px, 12px);

        :global(.uui-input-label) {
            margin-left: 6px;
        }
    }

    &.mode-cell {
        justify-content: center;
    }
}