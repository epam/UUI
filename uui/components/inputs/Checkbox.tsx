import { withMods } from '@epam/uui-core';
import * as uuiComponents from '@epam/uui-components';
import { ReactComponent as Check } from '@epam/assets/icons/notification-done-outline.svg';
import { ReactComponent as PartlySelect } from '@epam/assets/icons/content-minus-outline.svg';
import css from './Checkbox.module.scss';

interface CheckboxMods {
    /**
     * Defines component size.
     * @default '18'
     */
    size?: '12' | '18';
    /**
     * Defines the different edit modes.
     * @default 'form'
     */
    mode?: 'form' | 'cell';
}

/** Represents the properties of the Checkbox component. */
export type CheckboxProps = CheckboxMods & uuiComponents.CheckboxProps;

function applyCheckboxMods(mods: CheckboxMods) {
    return [
        css.root,
        css['size-' + (mods.size || '18')],
        css['mode-' + (mods.mode || 'form')],
        'uui-color-primary',
    ];
}

const applyUUICheckboxProps = (props: CheckboxProps) => {
    const defaultIcon = Check;
    const defaultIndeterminateIcon = PartlySelect;
    return {
        icon: props.icon ? props.icon : defaultIcon,
        indeterminateIcon: props.indeterminateIcon ? props.indeterminateIcon : defaultIndeterminateIcon,
    };
};

export const Checkbox = /* @__PURE__ */withMods<uuiComponents.CheckboxProps, CheckboxMods>(uuiComponents.Checkbox, applyCheckboxMods, applyUUICheckboxProps);
