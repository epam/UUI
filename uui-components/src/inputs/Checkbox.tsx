import * as React from 'react';
import { cx, uuiMarkers } from '@epam/uui';
import * as css from './Checkbox.scss';
import { Icon, uuiMod, uuiElement, isClickableChildClicked, CheckboxCoreProps, UuiContexts, UuiContext } from '@epam/uui';
import { IconContainer } from '../layout';

export interface CheckboxProps extends CheckboxCoreProps {
    icon?: Icon;
    indeterminateIcon?: Icon;
    renderLabel?(): React.ReactNode;
}

export class Checkbox extends React.Component<CheckboxProps, any> {
    static contextType = UuiContext;
    context: UuiContexts;

    handleChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
        !isClickableChildClicked(e) && this.props.onValueChange(!this.props.value);

        if (this.props.getValueChangeAnalyticsEvent) {
            const event = this.props.getValueChangeAnalyticsEvent(!this.props.value, this.props.value);
            this.context.uuiAnalytics.sendEvent(event);
        }
    }

    render() {
        return (
            <label
                role="checkbox"
                className={ cx(
                    css.container,
                    this.props.cx,
                    this.props.isDisabled && uuiMod.disabled,
                    this.props.isReadonly && uuiMod.readonly,
                    this.props.isInvalid && uuiMod.invalid,
                    (!this.props.isReadonly && !this.props.isDisabled) && uuiMarkers.clickable,
                ) }
                { ...this.props.rawProps }
            >
                <div className={ cx(uuiElement.checkbox, (this.props.value || this.props.indeterminate) && uuiMod.checked) }>
                    <input
                        type="checkbox"
                        onChange={ this.handleChange }
                        disabled={ this.props.isDisabled }
                        readOnly={ this.props.isReadonly }
                        aria-checked={ this.props.value }
                        checked={ this.props.value }
                    />
                    { this.props.value && !this.props.indeterminate && <IconContainer icon={ this.props.icon } /> }
                    { this.props.indeterminate && <IconContainer icon={ this.props.indeterminateIcon } /> }
                </div>
                { (this.props.renderLabel || this.props.label) && (
                    <div role="label" className={ uuiElement.inputLabel }>
                        { this.props.renderLabel ? this.props.renderLabel() : this.props.label }
                    </div>
                ) }
            </label>
        );
    }
}