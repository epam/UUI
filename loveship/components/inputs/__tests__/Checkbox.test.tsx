import React from 'react';
import renderer from 'react-test-renderer';
import { Checkbox } from '../Checkbox';

describe('Checkbox', () => {
    const value = true;
    const onChange = jest.fn();

    it('should be rendered correctly', () => {
        const tree = renderer.create(<Checkbox value={value} onValueChange={onChange} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with new props', () => {
        const tree = renderer.create(<Checkbox value={value} onValueChange={onChange} size={'12'} theme={'dark'} color="sun" />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
