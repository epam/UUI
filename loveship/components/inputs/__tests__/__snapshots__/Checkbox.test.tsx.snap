// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should be rendered correctly 1`] = `
<label
  className="container root size-18 theme-light color-sky -clickable"
  role="checkbox"
>
  <div
    className="uui-checkbox uui-checked"
  >
    <input
      aria-checked={true}
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    >
      <svg
        className=""
        height={24}
        viewBox="0 0 12 24"
        width={12}
      >
        <use
          xlinkHref="#checkbox-checked.svg"
        />
      </svg>
    </div>
  </div>
</label>
`;

exports[`Checkbox should be rendered correctly with new props 1`] = `
<label
  className="container root size-12 theme-dark color-sun -clickable"
  role="checkbox"
>
  <div
    className="uui-checkbox uui-checked"
  >
    <input
      aria-checked={true}
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    >
      <svg
        className=""
        height={24}
        viewBox="0 0 12 24"
        width={12}
      >
        <use
          xlinkHref="#checkbox-checked.svg"
        />
      </svg>
    </div>
  </div>
</label>
`;
